client.go

```diff
@@ -364,6 +364,48 @@ func (h hookLogger) Printf(s string, args ...interface{}) {
 	h.Info(fmt.Sprintf(s, args...))
 }
 
+// ContextLogger is an interface that provides methods for logging
+// with context. The methods accept a context.Context, a message
+// string and a variadic number of key-value pairs.
+type ContextLogger interface {
+	ErrorContext(ctx context.Context, msg string, keysAndValues ...interface{})
+	InfoContext(ctx context.Context, msg string, keysAndValues ...interface{})
+	DebugContext(ctx context.Context, msg string, keysAndValues ...interface{})
+	WarnContext(ctx context.Context, msg string, keysAndValues ...interface{})
+}
+
+// contextLogger adapts an ContextLogger to Logger for use by the existing hook functions
+// without changing the API.
+type contextLogger struct {
+	logger ContextLogger
+}
+
+// NewContextLogger returns a new contextLogger
+// for example: NewContextLogger(slog.Default())
+func NewContextLogger(logger ContextLogger) ContextLogger {
+	return &contextLogger{logger: logger}
+}
+
+func (s contextLogger) ErrorContext(ctx context.Context, msg string, keysAndValues ...interface{}) {
+	s.logger.ErrorContext(ctx, msg, keysAndValues...)
+}
+
+func (s contextLogger) InfoContext(ctx context.Context, msg string, keysAndValues ...interface{}) {
+	s.logger.InfoContext(ctx, msg, keysAndValues...)
+}
+
+func (s contextLogger) DebugContext(ctx context.Context, msg string, keysAndValues ...interface{}) {
+	s.logger.DebugContext(ctx, msg, keysAndValues...)
+}
+
+func (s contextLogger) WarnContext(ctx context.Context, msg string, keysAndValues ...interface{}) {
+	s.logger.WarnContext(ctx, msg, keysAndValues...)
+}
+
+func (s contextLogger) Printf(msg string, keysAndValues ...interface{}) {
+	s.logger.InfoContext(context.Background(), msg, keysAndValues...)
```

Comment:
I think `contextLogger` should have `ctx` as a field, which it then passes on here.

Reply:
Here is the method used to implement the Logger interface. It does not add the ctx parameter and maintains the original Printf function.

Reply:
But I am saying that it should. The whole point of this PR is that one can access logger and data from the context. And here you are passing `context.Background()` which breaks the whole point of this PR.